sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala

```diff
@@ -487,6 +487,88 @@ case class GpuInitCap(child: Expression) extends GpuUnaryExpression with Implici
     }
 }
 
+case class GpuStringRepeat(input: Expression, repeatTimes: Expression)
+    extends GpuBinaryExpression with ImplicitCastInputTypes with NullIntolerant {
+  override def left: Expression = input
+  override def right: Expression = repeatTimes
+  override def dataType: DataType = input.dataType
+  override def inputTypes: Seq[AbstractDataType] = Seq(StringType, IntegerType)
+
+  def doColumnar(input: GpuScalar, repeatTimes: GpuColumnVector): ColumnVector = {
+    assert(input.dataType == StringType)
+
+    withResource(GpuColumnVector.from(input, repeatTimes.getRowCount.asInstanceOf[Int],
+                 input.dataType)) {
+      replicatedInput => doColumnar(replicatedInput, repeatTimes)
+    }
+  }
+
+  def doColumnar(input: GpuColumnVector, repeatTimes: GpuColumnVector): ColumnVector = {
+    val repeatTimesCV = repeatTimes.getBase
+
+    // Compute the output size to check for overflow.
+    withResource(input.getBase.repeatStringsSizes(repeatTimesCV)) { outputSizes =>
+      if (outputSizes.getTotalSize > Int.MaxValue.asInstanceOf[Long]) {
+        throw new RuntimeException("Output strings have total size exceed maximum allowed size")
+      }
+
+      // Finally repeat the strings using the pre-computed strings' sizes.
+      input.getBase.repeatStrings(repeatTimesCV, outputSizes.getStringSizes)
+    }
+  }
+
+  def doColumnar(input: GpuColumnVector, repeatTimes: GpuScalar): ColumnVector = {
+    assert(repeatTimes.dataType == IntegerType)
+
+    if (!repeatTimes.isValid) {
+      // If the input scala repeatTimes is invalid, the results should be all nulls.
+      // This is equivalent to having repeatTimes == 0.
```

Comment:
Is this a true?  Looking at the Spark code, I would expect repeat times <= 0 to produce an _empty_ string, not a null string.

Reply:
Sorry I was wrong. Yes, repeat times <= 0 will result in empty strings. Will fix this.

Reply:
I thought one of tests would catch this.  Is that not happening?

Reply:
I have no idea. There is already unit test for that case (the last one below):

```def test_repeat_column_and_scalar():
    gen_s = StringGen(nullable=True)
    assert_gpu_and_cpu_are_equal_collect(
            lambda spark: unary_op_df(spark, gen_s).selectExpr(
                'repeat(a, -10)',
                'repeat(a, 0)',
                'repeat(a, 10)',
                'repeat(a, NULL)'
            ))
```

After fixing this (to return a column of nulls, instead of calling `repeat(a, 0`), the tests still passed all.

Reply:
Maybe that last unit test calls the `repeat(column, column)` version?

Reply:
Put in some debugging and find out.

Reply:
Maybe.  It's a red flag that you can change the behavior and tests that are supposed to detect that behavior don't notice.  Like @revans2 indicated, it's best to manually verify the tests do what you think they're going to do, otherwise we can end up writing redundant or useless tests.

It would be good to understand why the test that's supposed to catch this doesn't in practice and fix it.

Reply:
@revans2 @jlowe I added `throw new RuntimeException` into all my `doColumnar` functions. The following tests passed, which means these expressions are not executed by my code. So, these expressions are either ignored, or executed somewhere else:
```
def test_repeat_scalar_and_column():
    gen_r = IntegerGen(min_val=-100, max_val=100, special_cases=[0], nullable=True)
    assert_gpu_and_cpu_are_equal_collect(
            lambda spark: unary_op_df(spark, gen_r).selectExpr(
                'repeat(NULL, a)'
            ))

def test_repeat_column_and_scalar():
    gen_s = StringGen(nullable=True)
    assert_gpu_and_cpu_are_equal_collect(
            lambda spark: unary_op_df(spark, gen_s).selectExpr(
                'repeat(a, NULL)'
            ))
```

Reply:
BTW, is there any way to detect, and force NOT to, such compile-time evaluation of expressions?

Reply:
It shouldn't be runtime folded because StringRepeat is a regular binary op with

```
  override def foldable: Boolean = left.foldable && right.foldable
```

There is a way to disable it, but we guard against it and https://github.com/NVIDIA/spark-rapids/pull/1751/files makes it so we would fall back to the CPU in many of those cases.  The code there is a very special case where some times some of the code can return a constant so we need to cover it, but it does not need to be that fast.

Reply:
I'm not so sure this is solely a `foldable` thing.  `StringRepeat` is marked as `NullIntolerant` in Spark, and I wouldn't be surprised if the optimizer elided any subexpression marked `NullIntolerant` when any inputs can be computed at query planning time as null.

Reply:
The code sure looks like it is just about foldable.

https://github.com/apache/spark/blob/0f538402fb76e4d6182cc881219d53b5fdf73af1/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala#L41-L71

Reply:
`StringRepeat` in Spark is marked as `NullIntolerant` so I suspect the query optimizer eliminates any subexpression that it can tell at query planning time is going to receive a NULL.

I think we can make this work by adding a null column to the dataset and then performing a scalar subquery on it (e.g.: `SELECT repeat(stringscol, (SELECT first(nullcol) from dataset)) from dataset`), although I wouldn't be surprised if something like AQE ends up foiling the obfuscation.

Reply:
Yup it is a separate optimization. https://github.com/apache/spark/blob/0f538402fb76e4d6182cc881219d53b5fdf73af1/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala#L793-L794

I am not sure it is worth spending a lot of time hacking around this to make the test work, because it is highly unlikely that anyone will actually hit it. Who turns off these optimizations anyways.  But if you want to play games like what @jlowe is suggesting you can.

Reply:
> The code sure looks like it is just about foldable.

Yes, foldable can explain this if both sides are foldable, but even if they aren't (and in this case since one side is a column it's probably not foldable) `NullIntolerant` is where I think the trivial NULL projection is occurring.
```
/**
 * Replaces [[Expression Expressions]] that can be statically evaluated with
 * equivalent [[Literal]] values. This rule is more specific with
 * Null value propagation from bottom to top of the expression tree.
 */
object NullPropagation extends Rule[LogicalPlan] {
[..]
      // Non-leaf NullIntolerant expressions will return null, if at least one of its children is
      // a null literal.
      case e: NullIntolerant if e.children.exists(isNullLiteral) =>
        Literal.create(null, e.dataType)
    }
```